File: src/app/www/controllers/sort.rs
use std::{collections::HashMap, sync::Arc};

use axum::extract::{Query, State};
use maud::{html, Markup};

use crate::{
    app::www::{components::notification, I18n},
    domain::{models::Order, use_cases},
    infra::SalaryRepository,
};

use super::super::fragments::salary_table;

pub async fn get(
    State(repo): State<Arc<dyn SalaryRepository>>,
    Query(params): Query<HashMap<String, String>>,
) -> Markup {
    let order = Order::from(params);

    let salaries = match use_cases::fetch_salaries(repo, order.clone()).await {
        Ok(salaries) => salaries,
        Err(use_cases::fetch_salaries::Error::Unknown(_)) => {
            return html! { (notification::view(Some(I18n::SortError.translate()))) }
        }
    };

    salary_table::view(salaries, order)
}

File: src/app/www/controllers/sort.ts
import { Context } from "hono";
import { z } from "zod";
import { fetchSalaries } from "@domain";
import { SalaryRepository } from "@infra";
import { Maybe, Result } from "@utils";
import { SalaryTable } from "../fragments/mod.ts";
import { buildOrder, parseQuery } from "./utils/mod.ts";
import { Notification } from "../components/mod.ts";
import { I18n } from "../i18n.ts";

const querySchema = z.object({
    key: z.string().array().optional(),
    direction: z.string().array().optional(),
});

export async function get(
    c: Context,
    salaryRepo: SalaryRepository
) {
    const query = parseQuery(c, querySchema);
    const order = buildOrder(query);

    const salariesResult = await fetchSalaries(order, salaryRepo);
    if (Result.isErr(salariesResult)) { return c.html(Notification({
        notification: Maybe.some(I18n.translate("sort_error"))
    })); } 

    return c.html(SalaryTable({
        salaries: Result.unwrap(salariesResult),
        order
    }));
}